src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1045,95 +1075,202 @@ private static void ParseStatusLine(ReadOnlySpan<byte> line, HttpResponseMessage
             }
         }
 
-        private static void ParseHeaderNameValue(HttpConnection connection, ReadOnlySpan<byte> line, HttpResponseMessage response, bool isFromTrailer)
+        private bool ParseHeaders(HttpResponseMessage? response, bool isFromTrailer)
         {
-            Debug.Assert(line.Length > 0);
+            Span<byte> buffer = new Span<byte>(_readBuffer, _readOffset, _readLength - _readOffset);
 
-            int pos = 0;
-            while (line[pos] != (byte)':' && line[pos] != (byte)' ')
+            (bool finished, int bytesConsumed) = ParseHeadersCore(buffer, response, isFromTrailer);
+
+            int bytesScanned = finished ? bytesConsumed : buffer.Length;
+            if (_allowedReadLineBytes < bytesScanned)
             {
-                pos++;
-                if (pos == line.Length)
+                ThrowExceededAllowedReadLineBytes();
+            }
+
+            _readOffset += bytesConsumed;
+            _allowedReadLineBytes -= bytesConsumed;
+            Debug.Assert(_allowedReadLineBytes >= 0);
+
+            return finished;
+        }
+
+        private (bool finished, int bytesConsumed) ParseHeadersCore(Span<byte> buffer, HttpResponseMessage? response, bool isFromTrailer)
+        {
+            int originalBufferLength = buffer.Length;
+
+            while (true)
+            {
+                int colIdx = buffer.IndexOfAny((byte)':', (byte)'\n');
+                if (colIdx < 0) goto needMore;
```

Comment:
Can we `break;` instead of `goto` in all of these places?

Reply:
Replaced the 4 gotos with direct `return`s.
I'm not a fan of the duplication, but I'll measure if it has any impact of perf tomorrow.